public/js/src/module/controller-webform.js

```diff
@@ -258,12 +258,14 @@ function _submitRecord() {
 
     return fileManager.getCurrentFiles()
         .then( files => {
+            const csrfToken = ( document.cookie.split( '; ' ).find( c => c.startsWith( '__csrf' ) ) || '' ).split( '=' )[1];
```

Comment:
There is a [utility function](https://github.com/enketo/enketo-core/blob/master/src/js/utils.js#L125) in enketo-core that could perhaps be used as well with `import {readCookie} from 'enketo-core/src/js/utils';`

Reply:
Ah yes, I wanted to ask you about this. I saw it but got cold feet because, probably due to my rudimentary JS knowledge, I'm not confident that I understand the scope of `cookies`. Is it global? Does that mean that it gets set the first time that `readCookie` is used and then doesn't update even if new cookies are set? If that's right, it seems like it could lead to subtle bugs depending on where else it's called and the sequence of requests. Since I could write what I wanted as a relatively clear one-liner, that gave me more confidence. I also noticed that `setLogoutLinkVisibility` in the same file doesn't use the utility. But if you think it's the way to go, happy to do that.

Reply:
`cookies` is not global only accessible by utils.js, but indeed is only populated once upon load, which I think is fine because Enketo doesn't set any in-session cookies after loading and __csrf cookie is set before loading the form. Totally cool to keep as-is though.

Reply:
Thanks for being flexible and for that explanation. I guess it's the lifecycle of `utils` that I'm still not sure about. Does "upon load" mean upon loading Enketo frontend? So let's say I access an Enketo form link without auth, then `cookies` gets populated once. Then I get redirected to my data server, enter credentials and return. Does `cookies` gets populated again because Enketo client is reloaded?

Reply:
I meant when the HTML page is re-loaded (after having been redirected from Central's login which sets the cookie). Everything will start from scratch after that redirect, ie. the page is loaded, the scripts are loaded and the form is initialized. 

However, I now realize I was only thinking about the situation where Central returns a 401 at the time the form is loaded (in which case I believe we relocate to Central which sets the cookie, and redirects back to Enketo reloading the page (and reading the cookie that was just set). In case a 401 is received after initial form loading when the user is ready to submit a record she has been working on for hours (during which the authentication used for loading has expired), the page stays open and the user is prompted to open Central's login in a different tab. This would definitely not work if you had used readCookie from Enketo Core. I'm actually not sure if the current code works for this.

Reply:
> This would definitely not work if you had used readCookie from Enketo Core.

👍 Makes sense. I was thinking about the case of getting a 401 when an offline-able form gets connectivity and sends but maybe in that one it's actually ok because the browser would save the data? But yeah, it's clear that sending an online-only form and getting a 401 would be problematic.

> I'm actually not sure if the current code works for this.

Ah, thanks for the heads up. I'll take a look.

Reply:
@MartijnR I did some thinking through the case you mentioned and I think it works out ok. My understanding is:
1) the user clicks submit without having an auth cookie
1) the user sees a modal and clicks to go to the data server to auth in another tab (text generated in `controller-webform._submitRecord`)
1) the user is redirected back to the blank form in that tab
1) the user closes that new tab and sees the form they filled out
1) the user clicks submit and that **calls `connection.uploadRecord` again which rebuilds the request with the CSRF token and the auth cookie**

It's awkward because of the tab switching that a user may not know to do but I think that if they do get back to the correct tab they should be ok. Does that sound right?

I'd like to actually try this but unfortunately I'm struggling to do so locally because I don't know how to get Central and Enketo to talk not through the Docker config and because of the HTTPS requirements.

Reply:
Thanks for this. Yes for offline-capable forms, this is about correct I think. I'm just not sure if the cookie becomes available on that existing tab without refreshing. 

I think it's normal for an empty form to be displayed at the time a 401 is received, so the user could either go back to the original tab (if the cookies update without refresh) or continue working on the new tab. I think the use of `target="_blank"` was to cater to the very unlikely event that there is some unsaved data in the existing tab (and for online-only webform views).

Reply:
I can confirm that everything works beautifully for both offline and online forms in Chrome. The cookie becomes available immediately even if login happened in another tab.

Reply:
Great! :shipit: